File: sdk/typescript/src/contracts/risk-pool.ts
import {
  SorobanClient,
  TxOptions,
  toAddress,
  toI128,
  toU64,
  toU32,
  toBool,
  toString,
  fromScVal,
} from "../client";
import {
  InitializeRiskPoolSchema,
  DepositSchema,
  WithdrawSchema,
  SetVestingSchema,
  AllocateRewardsSchema,
  AppealSlashingSchema,
  ResolveAppealSchema,
  SetReinsuranceSchema,
  SetPauseStateSchema,
  type LiquidityProvider,
  type VestingConfig,
} from "../schemas";

export class RiskPoolContract {
  constructor(
    private readonly client: SorobanClient,
    private readonly contractId: string
  ) {}

  async initialize(
    admin: string,
    guardian: string,
    opts: TxOptions
  ): Promise<void> {
    const params = InitializeRiskPoolSchema.parse({ admin, guardian });
    await this.client.invoke(
      this.contractId,
      "initialize",
      [toAddress(params.admin), toAddress(params.guardian)],
      opts
    );
  }

  async deposit(
    from: string,
    amount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = DepositSchema.parse({ from, amount });
    await this.client.invoke(
      this.contractId,
      "deposit",
      [toAddress(params.from), toI128(params.amount)],
      opts
    );
  }

  async withdraw(
    to: string,
    amount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = WithdrawSchema.parse({ to, amount });
    await this.client.invoke(
      this.contractId,
      "withdraw",
      [toAddress(params.to), toI128(params.amount)],
      opts
    );
  }

  async setVestingParameters(
    caller: string,
    cliffSecs: bigint,
    durationSecs: bigint,
    penaltyBps: number,
    opts: TxOptions
  ): Promise<void> {
    const params = SetVestingSchema.parse({
      caller,
      cliff_secs: cliffSecs,
      duration_secs: durationSecs,
      penalty_bps: penaltyBps,
    });
    await this.client.invoke(
      this.contractId,
      "set_vesting_parameters",
      [
        toAddress(params.caller),
        toU64(params.cliff_secs),
        toU64(params.duration_secs),
        toU32(params.penalty_bps),
      ],
      opts
    );
  }

  async getVestingParameters(opts: TxOptions): Promise<VestingConfig> {
    const result = await this.client.invoke(
      this.contractId,
      "get_vesting_parameters",
      [],
      { ...opts, simulate: true }
    );
    return fromScVal<VestingConfig>(result);
  }

  async allocateRewards(
    caller: string,
    provider: string,
    amount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = AllocateRewardsSchema.parse({
      caller,
      provider,
      amount,
    });
    await this.client.invoke(
      this.contractId,
      "allocate_rewards",
      [
        toAddress(params.caller),
        toAddress(params.provider),
        toI128(params.amount),
      ],
      opts
    );
  }

  async claimVestedRewards(
    provider: string,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "claim_vested_rewards",
      [toAddress(provider)],
      opts
    );
  }

  async getProviderVestedRewards(
    provider: string,
    opts: TxOptions
  ): Promise<bigint> {
    const result = await this.client.invoke(
      this.contractId,
      "get_provider_vested_rewards",
      [toAddress(provider)],
      { ...opts, simulate: true }
    );
    return fromScVal<bigint>(result);
  }

  async getBalance(opts: TxOptions): Promise<bigint> {
    const result = await this.client.invoke(
      this.contractId,
      "get_balance",
      [],
      { ...opts, simulate: true }
    );
    return fromScVal<bigint>(result);
  }

  async appealSlashing(
    appealer: string,
    claimId: bigint,
    deposit: bigint,
    slashedAmount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = AppealSlashingSchema.parse({
      appealer,
      claim_id: claimId,
      deposit,
      slashed_amount: slashedAmount,
    });
    await this.client.invoke(
      this.contractId,
      "appeal_slashing",
      [
        toAddress(params.appealer),
        toI128(params.claim_id),
        toI128(params.deposit),
        toI128(params.slashed_amount),
      ],
      opts
    );
  }

  async resolveAppeal(
    caller: string,
    claimId: bigint,
    approved: boolean,
    refundPercentage: number,
    opts: TxOptions
  ): Promise<void> {
    const params = ResolveAppealSchema.parse({
      caller,
      claim_id: claimId,
      approved,
      refund_percentage: refundPercentage,
    });
    await this.client.invoke(
      this.contractId,
      "resolve_appeal",
      [
        toAddress(params.caller),
        toI128(params.claim_id),
        toBool(params.approved),
        toU32(params.refund_percentage),
      ],
      opts
    );
  }

  async setReinsuranceConfig(
    caller: string,
    reinsurer: string,
    percentage: number,
    creditScore: number,
    opts: TxOptions
  ): Promise<void> {
    const params = SetReinsuranceSchema.parse({
      caller,
      reinsurer,
      percentage,
      credit_score: creditScore,
    });
    await this.client.invoke(
      this.contractId,
      "set_reinsurance_config",
      [
        toAddress(params.caller),
        toAddress(params.reinsurer),
        toU32(params.percentage),
        toU32(params.credit_score),
      ],
      opts
    );
  }

  async cedeRisk(
    reinsurer: string,
    amount: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "cede_risk",
      [toAddress(reinsurer), toI128(amount)],
      opts
    );
  }

  async setPauseState(
    caller: string,
    isPaused: boolean,
    reason: string,
    opts: TxOptions
  ): Promise<void> {
    const params = SetPauseStateSchema.parse({
      caller,
      is_paused: isPaused,
      reason,
    });
    await this.client.invoke(
      this.contractId,
      "set_pause_state",
      [toAddress(params.caller), toBool(params.is_paused), toString(params.reason)],
      opts
    );
  }
}

File: sdk/python/stellar_insured_sdk/contracts/risk_pool.py
"""Risk Pool contract wrapper with schema-validated payloads."""

from __future__ import annotations

from typing import Any

from ..client import (
    SorobanClient,
    TxOptions,
    to_address,
    to_int128,
    to_uint64,
    to_uint32,
    to_bool,
    to_string,
    from_scval,
)
from ..schemas import (
    DepositParams,
    WithdrawParams,
    SetVestingParams,
    AllocateRewardsParams,
    AppealSlashingParams,
    ResolveAppealParams,
    SetReinsuranceParams,
    SetPauseStateParams,
)


class RiskPoolContract:
    """Typed wrapper for the Risk Pool Soroban contract."""

    def __init__(self, client: SorobanClient, contract_id: str) -> None:
        self._client = client
        self._contract_id = contract_id

    def initialize(self, admin: str, guardian: str, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "initialize",
            [to_address(admin), to_address(guardian)],
            opts,
        )

    def deposit(self, sender: str, amount: int, opts: TxOptions) -> None:
        params = DepositParams(**{"from": sender, "amount": amount})
        self._client.invoke(
            self._contract_id,
            "deposit",
            [to_address(params.sender), to_int128(params.amount)],
            opts,
        )

    def withdraw(self, to: str, amount: int, opts: TxOptions) -> None:
        params = WithdrawParams(to=to, amount=amount)
        self._client.invoke(
            self._contract_id,
            "withdraw",
            [to_address(params.to), to_int128(params.amount)],
            opts,
        )

    def set_vesting_parameters(
        self,
        caller: str,
        cliff_secs: int,
        duration_secs: int,
        penalty_bps: int,
        opts: TxOptions,
    ) -> None:
        params = SetVestingParams(
            caller=caller,
            cliff_secs=cliff_secs,
            duration_secs=duration_secs,
            penalty_bps=penalty_bps,
        )
        self._client.invoke(
            self._contract_id,
            "set_vesting_parameters",
            [
                to_address(params.caller),
                to_uint64(params.cliff_secs),
                to_uint64(params.duration_secs),
                to_uint32(params.penalty_bps),
            ],
            opts,
        )

    def get_vesting_parameters(self, opts: TxOptions) -> Any:
        result = self._client.invoke(
            self._contract_id,
            "get_vesting_parameters",
            [],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def allocate_rewards(
        self, caller: str, provider: str, amount: int, opts: TxOptions
    ) -> None:
        params = AllocateRewardsParams(caller=caller, provider=provider, amount=amount)
        self._client.invoke(
            self._contract_id,
            "allocate_rewards",
            [
                to_address(params.caller),
                to_address(params.provider),
                to_int128(params.amount),
            ],
            opts,
        )

    def claim_vested_rewards(self, provider: str, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "claim_vested_rewards",
            [to_address(provider)],
            opts,
        )

    def get_provider_vested_rewards(self, provider: str, opts: TxOptions) -> int:
        result = self._client.invoke(
            self._contract_id,
            "get_provider_vested_rewards",
            [to_address(provider)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def get_balance(self, opts: TxOptions) -> int:
        result = self._client.invoke(
            self._contract_id,
            "get_balance",
            [],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def appeal_slashing(
        self,
        appealer: str,
        claim_id: int,
        deposit: int,
        slashed_amount: int,
        opts: TxOptions,
    ) -> None:
        params = AppealSlashingParams(
            appealer=appealer,
            claim_id=claim_id,
            deposit=deposit,
            slashed_amount=slashed_amount,
        )
        self._client.invoke(
            self._contract_id,
            "appeal_slashing",
            [
                to_address(params.appealer),
                to_int128(params.claim_id),
                to_int128(params.deposit),
                to_int128(params.slashed_amount),
            ],
            opts,
        )

    def resolve_appeal(
        self,
        caller: str,
        claim_id: int,
        approved: bool,
        refund_percentage: int,
        opts: TxOptions,
    ) -> None:
        params = ResolveAppealParams(
            caller=caller,
            claim_id=claim_id,
            approved=approved,
            refund_percentage=refund_percentage,
        )
        self._client.invoke(
            self._contract_id,
            "resolve_appeal",
            [
                to_address(params.caller),
                to_int128(params.claim_id),
                to_bool(params.approved),
                to_uint32(params.refund_percentage),
            ],
            opts,
        )

    def set_reinsurance_config(
        self,
        caller: str,
        reinsurer: str,
        percentage: int,
        credit_score: int,
        opts: TxOptions,
    ) -> None:
        params = SetReinsuranceParams(
            caller=caller,
            reinsurer=reinsurer,
            percentage=percentage,
            credit_score=credit_score,
        )
        self._client.invoke(
            self._contract_id,
            "set_reinsurance_config",
            [
                to_address(params.caller),
                to_address(params.reinsurer),
                to_uint32(params.percentage),
                to_uint32(params.credit_score),
            ],
            opts,
        )

    def cede_risk(self, reinsurer: str, amount: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "cede_risk",
            [to_address(reinsurer), to_int128(amount)],
            opts,
        )

    def set_pause_state(
        self, caller: str, is_paused: bool, reason: str, opts: TxOptions
    ) -> None:
        params = SetPauseStateParams(
            caller=caller, is_paused=is_paused, reason=reason
        )
        self._client.invoke(
            self._contract_id,
            "set_pause_state",
            [
                to_address(params.caller),
                to_bool(params.is_paused),
                to_string(params.reason),
            ],
            opts,
        )
